[CI workflow] automate generating, configurating, and deploying build from `develop` branch (@W-19170135@)

vmarta opened the pull request:
This PR creates a daily workflow that generates (from `develop` branch), configures, and deploy a storefront build to MRT. 

Most of the challenges come from configuring the app after generating it. The simplest approach is to create a new generator preset that predefines the configuration values.

Related PR #3048 

## How to test drive the PR
We can truly test drive it once the PR is merged. But for now you can run this script:
- To see the generated app for performance tests: `node e2e/scripts/generate-project.js --project-key retail-react-app-performance-tests`
- For the demo site: `node e2e/scripts/generate-project.js --project-key retail-react-app-demo-site`

Afterwards, compare the following files:
- ssr.js and its contentSecurityPolicy
- config/default.js and its partialHydrationEnabled (PARTIAL_HYDRATION_ENABLED constant is now deprecated) 

The performance tests has its own unique needs, so you should see the values above different than the usual.

shethj committed:
Add new performance metrics workflow

shethj committed:
Add comment

vmarta committed:
Disable some steps that are not needed

vmarta committed:
Merge remote-tracking branch 'origin/develop' into feature/ci-pwa-kit-q3-perf

vmarta committed:
Update todos

vmarta committed:
Generate the project

vmarta committed:
Avoid dirtying workspace and remove unnecessary step

vmarta committed:
Tweak parameters

vmarta committed:
Allow for internal short codes

vmarta committed:
Align with the order of answers to the extensibility question

vmarta committed:
Create update-csp-directives.js

vmarta committed:
Rename script for clarity

vmarta committed:
Schedule it to run before the performance tests

vmarta committed:
Move constant into the config file

vmarta committed:
Move CSP directives into the config file

vmarta committed:
Make sure the corresponding .hbs files are up to date

vmarta committed:
Now scripts looks for the config file

vmarta committed:
Fix linting errors

vmarta committed:
Add auto lint fixing

vmarta committed:
For backward compatibility

vmarta committed:
New step to add CSP directives

vmarta committed:
Log where the file is at

vmarta committed:
Add a new script

vmarta committed:
Address a few todos

vmarta committed:
Allow updating property with the same value

vmarta committed:
Finish and clean up the workflow

vmarta committed:
Move scripts and remove default value for config path

vmarta committed:
No longer assume default value for config path

And no longer preserving comments

vmarta committed:
No longer support null, undefined values

vmarta committed:
Delete unnecessary comments

vmarta committed:
Revert changes in retail-react-app

vmarta committed:
Create a new generator preset

vmarta committed:
Revert changes to the generator

vmarta committed:
Add csp directives to ssr.js.hbs

vmarta committed:
Update app/constants.js files to support the new preset

vmarta committed:
No longer needing these scripts

vmarta committed:
Workflow now uses the new generator preset

vmarta committed:
No need to validate app in this case

vmarta committed:
Add the preset to more files

cc-prodsec commented:
### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/133d033c-bc17-46cc-a7de-433b29d88a7b?refs=5e48c542f19e1c2620a6dfd9041fd361af42a06c&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/133d033c-bc17-46cc-a7de-433b29d88a7b/license?refs=5e48c542f19e1c2620a6dfd9041fd361af42a06c&source=prChecksComment)

vmarta committed:
Revert some changes and keep one change

vmarta commented on .github/workflows/performance-metrics.yml:
```diff
@@ -0,0 +1,45 @@
+                  npm ci
+
+            - name: Generate App
+              uses: ./.github/actions/e2e_generate_app
+              with:
+                PROJECT_KEY: "retail-react-app-performance-tests"
```
Much simpler PR now because of creating this new generator preset.

vmarta commented on .github/actions/generate_app/action.yml:
```diff
@@ -31,20 +31,20 @@ runs:
       run: |
         use_extensibility_input="${{ inputs.use_extensibility }}"
         if [ "use_extensibility_input" = "true" ]; then
-         use_extensibility_value=2
-        else
          use_extensibility_value=1
+        else
+         use_extensibility_value=2
```
Bonus fix: the order of these options has changed, so we needed to swap the values.

vmarta committed:
Delete extra space

vmarta committed:
Merge branch 'develop' into vm/ci-pwa-kit-q3-perf-2

vmarta committed:
Empty commit

vmarta committed:
Simplify the handlebar template logic

vmarta committed:
No more hardcoding the `generated-projects` path

vmarta commented on .github/actions/e2e_generate_app/action.yml:
```diff
@@ -10,14 +10,24 @@ inputs:
       $COMMAND
+
+      # Return path to the generated project
+      GENERATED_PROJECTS_DIR=$(node -e "console.log(require('./e2e/config.js').GENERATED_PROJECTS_DIR)")
+      echo "project_path=$GENERATED_PROJECTS_DIR/${{ inputs.PROJECT_KEY }}" >> $GITHUB_OUTPUT
```
Bonus: outputting the path to generated project, so that other steps can easily use it and they don't need to hardcode the project path.

shethj commented on .github/workflows/e2e.yml:
```diff
@@ -134,7 +135,7 @@ jobs:
         uses: "./.github/actions/push_to_mrt"
         with:
-          CWD: "../generated-projects/retail-app-no-ext"
+          CWD: ${{ steps.generate_app_no_ext.outputs.project_path }}
```
Can you please manually trigger a run to the e2e tests on this branch to make sure nothing regressed since this PR modifies CWD in that workflow too ?

vmarta replied:
https://github.com/SalesforceCommerceCloud/pwa-kit/actions/runs/16913965897 looks good and passing.

bendvc commented on packages/pwa-kit-create-app/assets/templates/@salesforce/retail-react-app/app/constants.js.hbs:
I think it's best that we don't copy this entire file over. As per our discussion lets make a backwards compatible change where we move the island configure from the constants to the config file.

bendvc commented on packages/pwa-kit-create-app/program.json:
```diff
@@ -472,6 +472,56 @@
                 },
                 "private": true
             },
+            {
```
In an ideal world we'd use the new `--stdio` flag to pipe the configuration directly in to the generator. This means we don't have to change and release a new version of the create app project. But I think given the time constraint and the need for use to migrate this generator work to the `next` branch, and because you are making change in other parts of the generator, we can go ahead with this.

But in the future lets try to use that new feature, and ask ourselves if we see customers using these presets. If the answer is no, and we are only wanting one for some git hub actions, maybe we manage those externally.

vmarta committed:
Move feature toggle for partial hydration into the config file

vmarta committed:
Remove unnecessary tests

vmarta committed:
Update CHANGELOG.md

vmarta committed:
Make sure all ssr.js related files are in sync

vmarta committed:
Fix linting error

I wanted to preserve the comma at the end of the line because otherwise, I'm worried that it's easy to overlook it later.

vmarta commented on packages/template-retail-react-app/config/default.js:
```diff
@@ -60,7 +60,8 @@ module.exports = {
             tenantId: 'g82wgnrvm-ywk9dggrrw8mtggy.pc-rnd'
-        }
+        },
+        partialHydrationEnabled: false
```
As per Ben's suggestion, we now toggle the partial hydration feature via the config file (not the constants file).

vmarta commented on packages/template-retail-react-app/app/components/island/index.jsx:
```diff
@@ -84,7 +85,8 @@ function findChildren(children, componentType) {
 function Island(props) {
     const {children} = props
-    if (!PARTIAL_HYDRATION_ENABLED) {
+    const isEnabled = getConfig()?.app?.partialHydrationEnabled ?? PARTIAL_HYDRATION_ENABLED // in a backward compatible way
```
We also still support `PARTIAL_HYDRATION_ENABLED` for backward compatibility. But `config.app.partialHydrationEnabled` will take precedence if defined.

vmarta commented on packages/pwa-kit-create-app/assets/templates/@salesforce/retail-react-app/app/ssr.js.hbs:
```diff
@@ -299,11 +307,15 @@ const {handler} = runtime.createHandler(options, (app) => {
                 // Connect to Einstein APIs
                 'api.cquotient.com',
-                '*.c360a.salesforce.com'
+                // Connect to DataCloud APIs
+                '*.c360a.salesforce.com',
+                // Connect to SCRT2 URLs
+                '*.salesforce-scrt.com'
+            ],
+            'frame-src': [
+                // Allow frames from Salesforce site.com (Needed for MIAW)
+                '*.site.com'
```
Bonus: making sure that all ssr.js related files are in sync.
